pkg/spoke/submarineragent/config_controller.go

```diff
@@ -181,6 +191,49 @@ func (c *submarinerConfigController) sync(ctx context.Context, syncCtx factory.S
 		return c.updateGatewayStatus(syncCtx.Recorder(), config)
 	}
 
+	if config.Status.ManagedClusterInfo.Platform == "GCP" {
+		if meta.IsStatusConditionFalse(config.Status.Conditions, configv1alpha1.SubmarinerConfigConditionEnvPrepared) {
+			cloudProvider, preparedErr := cloud.GetCloudProvider(c.restMapper, c.kubeClient, nil, c.dynamicClient,
+				c.hubKubeClient, c.eventRecorder, config.Status.ManagedClusterInfo, config)
+
+			if preparedErr != nil {
+				return preparedErr
+			} else {
+				preparedErr = cloudProvider.PrepareSubmarinerClusterEnv()
+			}
+
+			condition := &metav1.Condition{
+				Type:    configv1alpha1.SubmarinerConfigConditionEnvPrepared,
+				Status:  metav1.ConditionTrue,
+				Reason:  "SubmarinerClusterEnvPrepared",
+				Message: "Submariner cluster environment was prepared",
+			}
+
+			if preparedErr != nil {
+				condition.Status = metav1.ConditionFalse
+				condition.Reason = "SubmarinerClusterEnvPreparationFailed"
+				condition.Message = fmt.Sprintf("Failed to prepare submariner cluster environment: %v", preparedErr)
+			}
+
+			managedClusterInfo := config.Status.ManagedClusterInfo
+			_, updated, updatedErr := helpers.UpdateSubmarinerConfigStatus(
+				c.configClient,
+				config.Namespace, config.Name,
+				helpers.UpdateSubmarinerConfigStatusFn(condition, managedClusterInfo),
+			)
+
+			if updatedErr != nil {
+				return err
+			}
+			if updated {
+				c.eventRecorder.Eventf("SubmarinerClusterEnvPrepared", "submariner cluster environment was prepared for manged cluster %s", config.Namespace)
+			}
+		}
+
+		updatedErr := c.updateGatewayStatus(syncCtx.Recorder(), config)
```

Comment:
@aswinsuryan @sridhargaddam @skitt Looking at this code in more detail, I notice that we call `updateGatewayStatus` on L191 above for AWS and return.  For GCP, we now also call `updateGatewayStatus` and return here so now we never get to `ensureGateways` below to actually label gateway nodes even for GCP. This doesn't seem right or am I missing something?

Reply:
@tpantelis  The gateway nodes are now labelled by cloud-prepare for AWS and GCP. The `ensureGateways` retained for other clouds like VMware and IBM cloud for tagging the node. In VMware/IBM cloud `SubmarinerAddon` does not program firewall rules and expect the user to do it.

Reply:
So it seems that if we have a cloud provider implementation for a platform, we can assume it's the cloud provider's responsibility to label the gateways?

 Another thing is the the hub controller checks if `config.Spec.CredentialsSecret` is nil otherwise it skips the cloud preparation.  It seems this should as well. I've added that check locally in the `cloud.ProviderFactory` added by https://github.com/open-cluster-management/submariner-addon/pull/149. However I wonder if we should the check in the spoke controller prior to code freeze.

Reply:
The managedClusterInfo will be created only config.Spec.CredentialsSecret is not nil in the hub. Spoke will respond only if managedClusterInfo is available. SO it should safe to assume config.Spec.CredentialsSecret will be set for GCP.

Reply:
> So it seems that if we have a cloud provider implementation for a platform, we can assume it's the cloud provider's responsibility to label the gateways?

Yes, cloud prepare should do that when we call ocpGatewayDeployer.Deploy()
> 
> Another thing is that the hub controller checks if `config.Spec.CredentialsSecret` is nil otherwise it skips the cloud preparation. It seems spoke controller should as well or is it safe to assume `config.Spec.CredentialsSecret` will be set for GCP? I've added that check in the `cloud.ProviderFactory` added by #149 however I'm sure if we should add the check in the spoke controller prior to code freeze to be safe.

Reply:
Right, it does AWS and GCP. Ideally this code shouldn't have knowledge of the internal implementation of specific cloud providers. We should just make gateway deployment a requirement and move the gateway labeling done here into a default cloud provider.

Reply:
> The managedClusterInfo will be created only config.Spec.CredentialsSecret is not nil in the hub. Spoke will respond only if managedClusterInfo is available. SO it should safe to assume config.Spec.CredentialsSecret will be set for GCP.

Actually the hub controller updates the `managedClusterInfo` regardless, ie even if `config.Spec.CredentialsSecret` is nil.

```
            if config.Spec.CredentialsSecret == nil {
		// no platform credentials, the cluster env is not requred to prepare, only update the manged cluster info
		_, _, err := helpers.UpdateSubmarinerConfigStatus(
			c.configClient,
			config.Namespace, config.Name,
			helpers.UpdateSubmarinerConfigStatusFn(&metav1.Condition{...}, managedClusterInfo),
		)
		return err
	}
```

But #149 is merged now.